src/Razor/test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Cohost/HtmlRequestInvokerTest.cs

```diff
@@ -0,0 +1,132 @@
+        lspDocumentManager.AddDocument(documentSnapshot.Uri, documentSnapshot);
+
+        var publisher = new TestHtmlDocumentPublisher();
+        var remoteServiceInvoker = new RemoteServiceInvoker();
+        var htmlDocumentSynchronizer = new HtmlDocumentSynchronizer(remoteServiceInvoker, publisher, LoggerFactory);
+        var invoker = new HtmlRequestInvoker(requestInvoker, lspDocumentManager, htmlDocumentSynchronizer, NoOpTelemetryReporter.Instance, LoggerFactory);
+
+        var validated = false;
+        requestInvoker.RequestAction = request =>
+        {
+            validated = true;
+            var diagnosticParams = Assert.IsType<VSInternalDiagnosticParams>(request);
+            Assert.Equal(htmlDocumentUri, diagnosticParams.TextDocument!.DocumentUri.GetRequiredParsedUri());
+        };
+
+        var diagnosticRequest = new VSInternalDiagnosticParams
+        {
+            TextDocument = new TextDocumentIdentifier { DocumentUri = document.CreateDocumentUri() }
+        };
+
+        _ = await invoker.MakeHtmlLspRequestAsync<VSInternalDiagnosticParams, object>(
+            document,
+            VSInternalMethods.DocumentPullDiagnosticName,
+            diagnosticRequest,
+            DisposalToken);
+
+        Assert.Equal(document.CreateDocumentUri(), diagnosticRequest.TextDocument!.DocumentUri);
+
+        Assert.True(validated);
+    }
+
+    [Fact]
+    public async Task ITextDocumentParamsRequest_UpdatesUri()
+    {
+        var document = Workspace.CurrentSolution.GetAdditionalDocument(_documentId).AssumeNotNull();
+
+        var htmlDocumentUri = new Uri("file://File.razor.html", UriKind.Absolute);
+
+        var htmlTextSnapshot = new StringTextSnapshot("");
+        var htmlTextBuffer = new TestTextBuffer(htmlTextSnapshot);
+        var checksum = await document.GetChecksumAsync(DisposalToken);
+        var requestInvoker = new TestLSPRequestInvoker((Methods.TextDocumentHoverName, null));
+        var lspDocumentManager = new TestDocumentManager();
+        var htmlVirtualDocument = new HtmlVirtualDocumentSnapshot(htmlDocumentUri, htmlTextBuffer.CurrentSnapshot, hostDocumentSyncVersion: 1, state: checksum);
+        var documentSnapshot = new TestLSPDocumentSnapshot(document.CreateUri(), version: (int)(htmlVirtualDocument.HostDocumentSyncVersion!.Value + 1), htmlVirtualDocument);
+        lspDocumentManager.AddDocument(documentSnapshot.Uri, documentSnapshot);
+
+        var publisher = new TestHtmlDocumentPublisher();
+        var remoteServiceInvoker = new RemoteServiceInvoker();
+        var htmlDocumentSynchronizer = new HtmlDocumentSynchronizer(remoteServiceInvoker, publisher, LoggerFactory);
+        var invoker = new HtmlRequestInvoker(requestInvoker, lspDocumentManager, htmlDocumentSynchronizer, NoOpTelemetryReporter.Instance, LoggerFactory);
```

Comment:
nit: Can we move all this boilerplate into a helper method that just returns something like the document/requestInvoker?